x/upgrade/keeper.go

```diff
@@ -169,15 +175,25 @@ func (k Keeper) TallyVotingPower(ctx sdk.Context, threshold int64) (bool, uint64
 
 // GetVotingPowerThreshold returns the voting power threshold required to
 // upgrade to a new version.
-func (k Keeper) GetVotingPowerThreshold(ctx sdk.Context) sdkmath.Int {
-	// contract: totalVotingPower should not exceed MaxUint64
+func (k Keeper) GetVotingPowerThreshold(ctx sdk.Context) (sdkmath.Int, error) {
 	totalVotingPower := k.stakingKeeper.GetLastTotalPower(ctx)
 	thresholdFraction := SignalThreshold(ctx.BlockHeader().Version.App)
-	product := totalVotingPower.MulRaw(thresholdFraction.Numerator)
-	if product.ModRaw(thresholdFraction.Denominator).IsZero() {
-		return product.QuoRaw(thresholdFraction.Denominator)
+	product, err := totalVotingPower.SafeMul(sdkmath.NewInt(thresholdFraction.Numerator))
+	if err != nil {
+		return sdkmath.ZeroInt(), err
+	}
+	res, err := product.SafeQuo(sdkmath.NewInt(thresholdFraction.Denominator))
+	if err != nil {
+		return sdkmath.ZeroInt(), err
+	}
+	mod, err := product.SafeMod(sdkmath.NewInt(thresholdFraction.Denominator))
+	if err != nil {
+		return sdkmath.ZeroInt(), err
+	}
+	if mod.IsZero() {
+		return res, nil
 	}
-	return product.QuoRaw(thresholdFraction.Denominator).AddRaw(1)
+	return res.AddRaw(1), nil
```

Comment:
This code seems complicated so I'm open to refactoring it if we're okay with dropping the `Fraction` type

Reply:
Found a way to refactor using Cosmos SDK decimal. See 3fdc1d7838338ce74f047c3347619b656b70cf59. Note this results in a small amount of precision loss when dealing with extremely large total voting powers but IMO the precision loss is tolerable in favor of cleaner code. Thoughts @cmwaters ?